### PR TITLE
Fix explorer item creation and disposal

### LINKED ISSUE #78153: Explorer: Can't create folder on newly created folder
1. Empty explorer
2. Click `New Folder` top action, name the folder and create it
3. Click `New Folder` again

Input box comes and goes.

### PR DESCRIPTION
**Pull request by jeanp413**
This PR reapplies the changes from #72627 and the improvements from [here](https://github.com/microsoft/vscode/pull/75695#issuecomment-503719205) and [here](https://github.com/microsoft/vscode/pull/75695#issuecomment-503900068).
Adittionaly has some changes to fix #78153 reported issues

### CODE REVIEW

File: src/vs/base/browser/ui/tree/indexTreeModel.ts [hunk 1]
```diff
@@ -219,7 +219,7 @@ export class IndexTreeModel<T extends Exclude<any, undefined>, TFilterData = voi
 
 		const result = this._setListNodeCollapsed(node, listIndex, revealed, collapsed!, recursive || false);
 
-		if (this.autoExpandSingleChildren && !collapsed! && !recursive) {
+		if (node !== this.root && this.autoExpandSingleChildren && !collapsed! && !recursive) {
```
Comment (jeanp413) [inline src/vs/base/browser/ui/tree/indexTreeModel.ts]: This fixes the bug from the second gif from https://github.com/microsoft/vscode/issues/78153#issuecomment-522271070

Comment (isidorn) [reply src/vs/base/browser/ui/tree/indexTreeModel.ts]: Can you please put this in a seperate PR. I like this simple change and it is in the tree land so I would love if we merge it in unrelated to this.

File: src/vs/workbench/contrib/files/browser/views/explorerView.ts [hunk 2]
```diff
@@ -335,6 +343,13 @@ export class ExplorerView extends ViewletPanel {
 
 		this._register(this.tree.onContextMenu(e => this.onContextMenu(e)));
 
+		this._register(this.tree.onDidScroll(e => {
+			let editable = this.explorerService.getEditable();
```
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/views/explorerView.ts]: Why not use `getEditableData`?
That API is already there and gives you everything you need.

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
@jeanp413 thanks a lot for your PR.
I have added some comments inline, can you please address them?

Also did you intensivly test this out? There are quite some corner cases (as you are probably aware of). And this code is quite delicate (hacky).
I tried it out quickly and seems to work well (but plan to test it more once we adress the things I mentioned in the review).

File: src/vs/workbench/contrib/files/browser/views/explorerView.ts (hunk 2 above)
Comment (jeanp413) [reply src/vs/workbench/contrib/files/browser/views/explorerView.ts]: `getEditableData` just return the data, I need the stat so I can call `this.tree.getRelativeTop(editable.stat)`

File: src/vs/base/browser/ui/tree/indexTreeModel.ts (hunk 1 above)
Comment (jeanp413) [reply src/vs/base/browser/ui/tree/indexTreeModel.ts]: Ok

Comment (jeanp413) [reply src/vs/base/browser/ui/tree/indexTreeModel.ts]: Opened #79540 and it's corresponding PR #79542

**Comment by jeanp413** [comment]
I tested it for the bugs reported in #72626, #75693. #75624, #78153
Did you find a corner case that's not covered?

**Comment by isidorn** [comment]
@jeanp413 thanks for the PR and for updating it.
However for now I have decided to go with the current approach. And we can change this next milestone after we get some feedback, thus assigning it to september. The current limitations of the approach are dangling input elements and the input box not closing on scroll. For now I can live with that.

**Comment by isidorn** [comment]
I have tested this now more. Seems to work fine and looks good. Let's merge it in to get feedback from insiders users.
Thanks a lot for this PR.

**Pull request merged**